### ISSUE TITLE
Schema validation errors can not be captured by customized error handler

### ISSUE THREAD

**Issue by BigMurry** [fastify/fastify#341]
I've setup an error handler by using `setErrorHandler` method. But the schema validation errors can not be captured by my customized error handler.  Does it meant to act this way?

https://github.com/fastify/fastify/blob/1512439e45a08dbe4ed3854dacf4c06c733d7017/lib/handleRequest.js#L132-L140

From this code snippet, the `store` object for `Reply` constructor is `null`.

**Comment by mcollina** [comment]
Can you please provide an example to reproduce? What is the current output vs what would you expect?

**Comment by BigMurry** [comment]
First register the errorHandler:
```js
fastify.setSchemaCompiler(schema => data => Joi.validate(data, schema))
fastify.setErrorHandler(function (err, reply) {
  console.log('aaaa') // never get called when schema validation error occurred
  reply.code(err.statusCode).send(err)
})
const opts = {
  schema: {
    body: Joi.object().keys({
      a: Joi.string().valid('A', 'B')
    }).required()
  }
}
fastify.post('/test', opts, function (request, reply) { ... })
```

Then when I `POST /test` with body `{a: 'C'}`, I get:
```js
{
    "error": "Bad Request",
    "message": "child \"a\" fails because [\"a\" must be one of [A, B]]",
    "statusCode": 400
}

```
but the `console.log('aaaa')` never get called.

**Comment by mcollina** [comment]
Do you know or other people know how Hapi handle this situation?

**Comment by StarpTech** [comment]
@mcollina in hapi you can define a `onPreResponse`

```js
var onPreResponse = function(request, reply) {
  var response = request.response;

  if (response && response.isBoom) {
    response.output.message = 'Generic Error Message';
  }

  reply.continue();
}
```

**Comment by StarpTech** [comment]
@mcollina the store is `null` [here](https://github.com/fastify/fastify/blob/master/lib/reply.js#L166) because [here](https://github.com/fastify/fastify/blob/master/lib/handleRequest.js#L133)

the custom error handler is behind `reply.res._context.errorHandler`


this would fix it but I don't know about all consequences

```diff
function wrapReplyEnd (req, res, statusCode, payload) {
+ const reply = new Reply(req, res, res._context)
  if (payload instanceof Error) {
    reply.code(statusCode).send(payload)
  } else {
    reply.code(statusCode).send(new Error(payload || ''))
  }
  return
}
```

**Comment by mcollina** [comment]
Go ahead and fire a PR. Let’s discuss it there.

**Comment by StarpTech** [comment]
Already addressed in #330

**Comment by BigMurry** [comment]
Also there is another problem here. 
The `extendServerError` decorator will have no use here, cause the `extendServerError` decorator is behind `store.Reply.prototype._extendServerError`. 
And this would fix both the customized error handler and extendServerError decorator issue.
```diff
- function wrapReplyEnd (req, res, statusCode, payload) {
+ function wrapReplyEnd (store, req, res, statusCode, payload) {
- const reply = new Reply(req, res, null)
+ const reply = new store.Reply(req, res, store)
  if (payload instanceof Error) {
    reply.code(statusCode).send(payload)
  } else {
    reply.code(statusCode).send(new Error(payload || ''))
  }
  return
}
```

Good news, already addressed in #330.

**Comment by BigMurry** [comment]
close this as fixed in #330.
